static/js/containers/pages/checkout/CartPage.js

```diff
@@ -0,0 +1,276 @@
+// @flow
+/* global SETTINGS: false */
+import React from "react"
+import DocumentTitle from "react-document-title"
+import { CART_DISPLAY_PAGE_TITLE } from "../../../constants"
+import { compose } from "redux"
+import { connect } from "react-redux"
+import { connectRequest } from "redux-query"
+import type { BasketItem } from "../../../flow/cartTypes"
+
+import Loader from "../../../components/Loader"
+
+import { createStructuredSelector } from "reselect"
+import { cartQuery, cartSelector } from "../../../lib/queries/cart"
+
+import type { RouterHistory } from "react-router"
+import moment from "moment"
+import { formatPrettyDateTimeAmPmTz, parseDateString } from "../../../lib/util"
+import { Button } from "reactstrap"
+
+type Props = {
+  history: RouterHistory,
+  cartItems: Array<BasketItem>,
+  isLoading: boolean
+}
+
+export class CartPage extends React.Component<Props> {
+  renderCartItemCard(cartItem: BasketItem) {
+    const purchasableObject = cartItem.product.purchasable_object
```

Comment:
One thing I discovered in my testing is that it's possible for `purchaseable_object` to be `null` because Django's generic object mechanism doesn't enforce referenctial integrity at the DB level, which causes this page to error on render. We probably need to guard against such products being added to the cart, but even with that we probably should have this case.

The options I can think of are:

a) filter out cart items w/ a product that has no `purchasable_object`, the frontend will then treat it like an empty cart
b) the frontend checks for this kind of state and displays an error message

I'm fine with either of these options, but I'm not sure if there are broader implications here too.

Reply:
If I understand this correctly, the user tried to put something in the cart that isn't (currently?) purchasable? In a case like that, I think an error message is called for. 

Could be as simple as "Something went wrong. Contact Customer Support" for now.

Reply:
Planning on going with option A here - ideally products with no purchasble_object should get filtered out on the frontend anyway, before the checkout process, since you won't be able to activate the Enroll button for courses that aren't enrollable anyway. 

(There's an edge case here, though: if you enroll on the cusp of the window closing, you could have a cart item that was valid when you start but isn't once you hit the Place Order button. Could take care of this with a grace period but that can maybe be a task for later as it feels like this state wouldn't happen particularly often. If the cart ends up empty for that reason, it can pop up a "please contact support" message.)